usmqe/api/tendrlapi/common.py

```diff
@@ -1,20 +1,94 @@
-"""
-Tendrl REST API common functions.
-"""
+"""Tendrl REST API common functions."""
 
+import json
 import time
+
 import pytest
 import requests
+
 from usmqe.api.base import ApiBase
 
 LOGGER = pytest.get_logger("commonapi", module=True)
 
 
+class TendrlAuth(requests.auth.AuthBase):
+    """
+    Implementation of Tendrl Auth Method (Bearer Token) for requests
+    library, based on upstream documentation:
+
+    https://github.com/Tendrl/api/blob/master/docs/authentication.adoc
+
+    This way, auth. implementation is stored in a single place and it also
+    makes possible to omit auth object entrirely (thanks to design of
+    requests library) for testing purposes (eg. checks of corner cases to make
+    sure that security is not compromised) no matter if it makes sense from
+    Tendrl user perspective.
+
+    See also: http://docs.python-requests.org/en/master/user/authentication/
+    """
+
+    def __init__(self, token, username=None):
+        self._bearer_token = token
+        # metadata attributes for easier debugging, we need to trust login
+        # function to store correct values there
+        self.username = username
+
+    def __call__(self, r):
+        """Add Tendl Bearer Token into header of the request."""
+        headers = {
+            "Authorization": "Bearer {}".format(self._bearer_token),
+            }
+        r.prepare_headers(headers)
+        return r
+
+
 class TendrlApi(ApiBase):
     """ Common methods for Tendrl REST API.
     """
 
-    def get_job_attribute(self, job_id, attribute="status", section=None):
+    def login(self, username, password, asserts_in=None):
+        """ Login user.
+
+        Name:        "login",
+        Method:      "GET",
+        Pattern:     "login",
+
+        Args:
+            username: name of user that is going logged in
+            password: password for username
+            asserts_in: assert values for this call and this method
+        """
+        pattern = "login"
+        post_data = {"username": username, "password": password}
+        request = requests.post(
+            pytest.config.getini("usm_api_url") + pattern,
+            data=json.dumps(post_data))
+        self.print_req_info(request)
+        self.check_response(request, asserts_in)
+        LOGGER.debug("access_token: {}".format(request.json()))
+        token = request.json().get("access_token")
+        auth = TendrlAuth(token, username)
+        return auth
+
+    def logout(self, asserts_in=None, auth=None):
```

Comment:
About `auth=None` I agree that auth doesn't have to be mandatory parameter so we could easily test the case when the authentication is missing. On the other hand we could test that with calling the method with `auth=None` and moreover this one is almost only example when the auth is set like that - it is not mandatory. The other one is `get_job_attribute`. Please unify the approach in all methods to be mandatory or not.